contributors/design-proposals/core-metrics-pipeline.md

```diff
@@ -0,0 +1,168 @@
+type MemoryUsage struct {  
+  // The time at which these metrics were updated.  
+  Timestamp metav1.Time  
+  // The amount of "working set" memory. This includes recently accessed memory,  
+  // dirty memory, and kernel memory.  
+  UsageBytes *uint64  
+}   
+
+// FilesystemUsage holds statistics about the quantity of local storage (e.g. disk) resources consumed  
+type FilesystemUsage struct {  
+  // The time at which these metrics were updated.  
+  Timestamp metav1.Time  
+  // This must uniquely identify the storage resource that is consumed.
+  StorageIdentifier string  
+  // UsedBytes represents the disk space consumed, in bytes.  
+  UsedBytes *uint64  
+  // UsedInodes represents the inodes consumed  
+  UsedInodes *uint64  
+}  
+```
+
+### On-Demand Design
+Interface:  
+The interface for exposing these metrics within the kubelet contains methods for fetching each relevant metric.  These methods contains a "recency" parameter which specifies how recently the metrics must have been computed.  Kubelet components which require very up-to-date metrics (eviction, for example), use very low values.  Other components use higher values.  
+
+Implementation:  
+To keep performance bounded while still offering metrics "On-Demand", all calls to get metrics are cached, and a minimum recency is established to prevent repeated metrics computation.  Before computing new metrics, the previous metrics are checked to see if they meet the recency requirements of the caller.  If the age of the metrics meet the recency requirements, then the cached metrics are returned.  If not, then new metrics are computed and cached.  
+
+## Implementation Plan 
+@dashpole will modify the structure of metrics collection code to be "On-Demand".   
+
+Suggested, tentative future work, which may be covered by future proposals:  
+ - Publish these metrics in some form to a kubelet API endpoint
+ - Obtain all runtime-specific information needed to collect metrics from the CRI.   
+ - Kubernetes can be configured to run a default "third party metrics provider" as a daemonset.  Possibly standalone cAdvisor.
+
+## Rollout Plan
+Once this set of metrics is accepted, @dashpole will begin discussions on the format, and design of the endpoint that exposes them.  The node resource metrics endpoint (TBD) will be added alongside the current Summary API in an upcoming release.  This should allow concurrent developments of other portions of the system metrics pipeline (metrics-server, for example).  Once this addition is made, all other changes will be internal, and will not require any API changes.  
+@dashpole will also start discussions on integrating with the CRI, and discussions on how to provide an out-of-the-box solution for the "third party monitoring" pipeline on the node.  One current idea is a standalone verison of cAdvisor, but any third party metrics solution could serve this function as well.
+
+## Implementation Status
```

Comment:
nit: I'd get rid of this section. Proposals aren't the best medium for tracking status, and tend to get out of date. I'd use a [feature issue](https://github.com/kubernetes/features/issues/new) instead.